src/index.ts

```diff
@@ -1,122 +1,112 @@
-            ActionCreator<P3>
+            ActionCreator<P3>,
         ],
         handler: Handler<InS, OutS, P1 | P2 | P3>,
-    ): ReducerBuilder<InS, OutS>;
+    ): ReducerBuilder<InS, OutS, PassedS>;
     cases<P1, P2, P3, P4>(
         actionCreators: [
             ActionCreator<P1>,
             ActionCreator<P2>,
             ActionCreator<P3>,
-            ActionCreator<P4>
+            ActionCreator<P4>,
         ],
         handler: Handler<InS, OutS, P1 | P2 | P3 | P4>,
-    ): ReducerBuilder<InS, OutS>;
+    ): ReducerBuilder<InS, OutS, PassedS>;
     cases<P>(
         actionCreators: Array<ActionCreator<P>>,
         handler: Handler<InS, OutS, P>,
-    ): ReducerBuilder<InS, OutS>;
+    ): ReducerBuilder<InS, OutS, PassedS>;
 
     // casesWithAction variadic overloads
-    casesWithAction<P1>(
-        actionCreators: [ActionCreator<P1>],
-        handler: Handler<InS, OutS, Action<P1>>,
-    ): ReducerBuilder<InS, OutS>;
     casesWithAction<P1, P2>(
         actionCreators: [ActionCreator<P1>, ActionCreator<P2>],
         handler: Handler<InS, OutS, Action<P1 | P2>>,
-    ): ReducerBuilder<InS, OutS>;
+    ): ReducerBuilder<InS, OutS, PassedS>;
     casesWithAction<P1, P2, P3>(
         actionCreators: [
             ActionCreator<P1>,
             ActionCreator<P2>,
-            ActionCreator<P3>
+            ActionCreator<P3>,
         ],
         handler: Handler<InS, OutS, Action<P1 | P2 | P3>>,
-    ): ReducerBuilder<InS, OutS>;
+    ): ReducerBuilder<InS, OutS, PassedS>;
     casesWithAction<P1, P2, P3, P4>(
         actionCreators: [
             ActionCreator<P1>,
             ActionCreator<P2>,
             ActionCreator<P3>,
-            ActionCreator<P4>
+            ActionCreator<P4>,
         ],
         handler: Handler<InS, OutS, Action<P1 | P2 | P3 | P4>>,
-    ): ReducerBuilder<InS, OutS>;
+    ): ReducerBuilder<InS, OutS, PassedS>;
     casesWithAction<P>(
         actionCreators: Array<ActionCreator<P>>,
         handler: Handler<InS, OutS, Action<P>>,
-    ): ReducerBuilder<InS, OutS>;
+    ): ReducerBuilder<InS, OutS, PassedS>;
 
     withHandling(
         updateBuilder: (
-            builder: ReducerBuilder<InS, OutS>,
-        ) => ReducerBuilder<InS, OutS>,
-    ): ReducerBuilder<InS, OutS>;
+            builder: ReducerBuilder<InS, OutS, PassedS>,
+        ) => ReducerBuilder<InS, OutS, PassedS>,
+    ): ReducerBuilder<InS, OutS, PassedS>;
 
     // Intentionally avoid AnyAction in return type so packages can export reducers
     // created using .default() or .build() without requiring a dependency on typescript-fsa.
     default(
         defaultHandler: Handler<InS, OutS, AnyAction>,
-    ): (state: InS | undefined, action: { type: any }) => OutS;
-    build(): (state: InS | undefined, action: { type: any }) => OutS;
-    (state: InS | undefined, action: AnyAction): OutS;
+    ): (state: PassedS, action: AnyAction) => OutS;
+    build(): (state: PassedS, action: AnyAction) => OutS;
```

Comment:
As noted in the comment a few lines up, `AnyAction` was intentionally avoided here.

(Reading it I realize it's a bit vaguely phrased. It should say "without **consumers** requiring a dependency on typescript-fsa")